src/v/storage/disk_log_impl.cc

```diff
@@ -1261,6 +1261,15 @@ ss::future<> disk_log_impl::apply_segment_ms() {
         co_return;
     }
     auto last = _segs.back();
+
+    // Holding the append lock blocks writes to the last open segment.
+    // This is required in order to avoid the logic in this function
+    // racing with an inflight append. Contention on this lock should
+    // be very light, since we wouldn't need to enforce segment.ms
+    // if this partition was high throughput (segment would have rolled
+    // naturally).
+    auto append_lock_holder = co_await last->append_lock();
```

Comment:
I think I understand why we're doing it up here, but generally it seems like it would be safer to do this in release appender or to make sure other callers to release_appender take the append lock. No explicit races come to mind, but if thinking about this lock as protecting access to `_appender` or protecting swaps of `_appender`, seems like we should be covering all such cases.

Reply:
I think your suggestion would also work. However, specifically for `segment.ms` application, if we were to take the lock in `release_appender`, we might miss the fact that the in-flight write has rolled the segment while we were waiting for the append lock. This will lead to having a segment with a single data batch on disk. Granted, that I'm pretty sure this can happen on `dev` now too.

More generally, I don't think of the append lock as "protecting access to the `_appender`". To me that's an implementation detail. Instead, I see the semantics as "block appends while the lock is held". Callers should figure out if they have this requirement. For instance, when the segment rolls naturally via the write path (the only other place where we call into `release_appender`; see `disk_log_appender::operator()`), the roll and append are performed sequentially. There the caller doesn't grab the append lock because it doesn't need to block appends at any point.

Reply:
Yea there's a conceptual elegance to having this block appends, and I like the idea of having them be decoupled. But I worry that they're currently not decoupled enough.

At least here, we take the append lock and then take the read/write lock when we release.
On the append path, we take the read lock first (in `disk_log_appender::initialize`) and then take the append lock.

That smells it maybe could deadlock, because read locks can get stuck behind write locks, and these two facts point to lock ordering inversion. OTOH, I'm having to contort a bit to get there with regards to ordering of possible operations. Still, worth calling out and thinking about. Is there an invariant with segment usage that prevents this from being a problem?

Reply:
I think we do end up avoiding a deadlock. I was worried we would end up with a deadlock when append races with a roll:
t1: take read lock (append path)
t2: take append lock (release path)
t2: try to take the write lock
t2: wait indefinitely because t1 has read lock

We're safe from this because we first try the write lock before attempting to acquire it, so what actually would happen:
t1: take read lock (append path)
t2: take append lock (release path)
t2: fail to take write lock
t2: take read lock
t2: drop append lock
t2: wait to take write lock in background
t1: take append lock
t1: begin to append
t2: close appender

As far as deadlocks are concerned, i think we're okay, but I do think the segfault i'm seeing is that the appender gets swapped out right after the `t2: take read lock` step, before `t1: begin to append`